### DISCUSSION TITLE
Experiencing a weird error with Nextjs 15

### DISCUSSION THREAD

**Discussion by popovidis** [pocketbase/pocketbase#5843] [Q&A] [answered]
Hi there, I have been using pocketbase for close to a year now and I have reused all of my code that I made initially, so like simple utility functions, etc. 

I am having an issue specifically with this function:

```
export const getAdminPocketbase = async () => {
  pb.autoCancellation(false);
  await pb.admins.authWithPassword(PB_EMAIL, PB_PASSWORD);
  return pb;
};
```

On some pages it works, on others it does not...

Here is an example where it does not work

```
const Blog = async () => {
  const getBlogs = async () => {
    "use server";
    const pb = await getAdminPocketbase();
    return await pb.collection("blogs").getFullList({ filter: `body!=null` });
  };

  const blogs = await getBlogs();

  const featured = blogs.find((blog) => blog.featured);

  return ...
 ```
 
 Here is an example of it working:
 
 ```
 "use server"
 
 import ...
 
 export const getBlogPost = async (slug) => {
  try {
    const pb = await getAdminPocketbase();
    const blogPost = await pb
      .collection("blogs")
      .getFirstListItem(`slug="${slug}"`);
    if (blogPost) {
      return { success: true, data: blogPost };
    } else {
      return { success: false };
    } catch (e) {...}}
  ```
  
  When I try to console.log(pb) in the function it logs the correct info only on the second usage case. On the first it gives this error:
  
  ```
   ⨯ TypeError: Converting circular structure to JSON
    --> starting at object with constructor 'Client'
    |     property 'admins' -> object with constructor 'AdminService'
    --- property 'client' closes the circle
    at stringify (<anonymous>)
    at getAdminPocketbase (/home/nikolay/.code/bravobot/nextjs/.next/server/chunks/ssr/_c813b9._.js:304:13)
    at getBlogs (/home/nikolay/.code/bravobot/nextjs/.next/server/chunks/ssr/_c813b9._.js:590:163)
    at Blog (/home/nikolay/.code/bravobot/nextjs/.next/server/chunks/ssr/_c813b9._.js:597:25)
    at AsyncLocalStorage.run (node:async_hooks:346:14)
    at stringify (<anonymous>)
    at AsyncLocalStorage.run (node:async_hooks:346:14)
    at AsyncResource.runInAsyncScope (node:async_hooks:206:9)
digest: "602441042"
```

This is reused code from nextjs 14, not sure what changed to break this.

**Comment by ganigeorgiev** [comment]
The PocketBase JS SDK instance contains circular dependencies and it seems that you are trying to serialize it as json and it fails (_this is somewhat expected_).
What triggers the serialization I don't know (_it could be runtime specific behavior of `console.log(pb)` or some other part of your code_).

One thing to note though is that if `getAdminPocketbase()` is called on every request then this would be very slow because of the bcrypt password hash verification.

The above code doesn't show where the PocketBase client is actually created and I don't understand why do you have `getAdminPocketbase()` separate from it but keep in mind "Problem 1" mentioned in https://github.com/pocketbase/pocketbase/discussions/5313.

**Reply by popovidis** [reply]
This is what I have

```
import PocketBase from "pocketbase";
import { PB_EMAIL, PB_PASSWORD, PB_URL } from "./constants";

export const pb = new PocketBase(PB_URL);

export const getPocketBase = (cookie) => {
  if (cookie) {
    pb.authStore.loadFromCookie(cookie);
  }
  return pb;
};

export const getAdminPocketbase = async () => {
  pb.autoCancellation(false);
  await pb.admins.authWithPassword(PB_EMAIL, PB_PASSWORD);
  return pb;
};
```

Then I just call it when and where I need it from this file.

**Reply by ganigeorgiev** [reply]
Again, make sure to read "Problem 1" in https://github.com/pocketbase/pocketbase/discussions/5313.

The above code would be vulnerable if your Next.js server runs as long running node process because requests from different users will write over the same `pb.authStore` object.

**Reply by popovidis** [reply]
So the solution would be to call for a new PocketBase instance in every function where I need to verify users auth?

**Reply by ganigeorgiev** [reply]
In general - Yes, you need to call `new PocketBase` for each request.

A global PocketBase instance is fine for the admin-only use case but you can't mix that instance with the one for the regular users auth.

If you find this confusing then I'm sorry. I can't explain it better then what is already written in https://github.com/pocketbase/pocketbase/discussions/5313 and I'm tired of seeing buggy PocketBase applications caused by it. JS SSR comes with a lot of caveats when used with external auth solutions like PocketBase (Next.js makes things even more difficult) and users need to carefully evaluate whether the cost is worth it.

**Reply by popovidis** [reply]
```
export const getPocketBase = (cookie) => {
  const pb = new PocketBase(PB_URL);
  if (cookie) {
    pb.authStore.loadFromCookie(cookie);
  }
  return pb;
};

export const getAdminPocketbase = async () => {
  const pb = new PocketBase(PB_URL);
  pb.autoCancellation(false);
  await pb.admins.authWithPassword(PB_EMAIL, PB_PASSWORD);
  return pb;
};
```

Is this better?

Also any idea how to solve the problem I am facing?

**Reply by ganigeorgiev** [reply, marked as answer]
> Is this better?

Yes, but as mentioned in the beginning calling `authWithPassword` for every request would be very slow. It is safe to have a global admin client and reuse the token state once authenticated. Check the example at the end in https://github.com/pocketbase/pocketbase/discussions/5313.

> Also any idea how to solve the problem I am facing?

It is not clear what actually is the problem. The error message indicates that it doesn't allow calling `console.log(pb)` and as mentioned initially this is somewhat expected because the serialization of the `pb` object will vary from one runtime to another (whatever runtime you are using it seems that `console.log(pb)` internally is doing something like `console.log(JSON.stringify(pb))` to print the object). So in a sense to solve it don't call `console.log(pb)`.

**Reply by popovidis** [reply]
Weird alright. Thank you for helping me improve my functionality also!

**Reply by popovidis** [reply]
Finally, just as a review by the creator, hows this?

```
const adminPb = new PocketBase(PB_URL);

await adminPb.admins.authWithPassword(PB_EMAIL, PB_PASSWORD);
adminPb.autoCancellation(false);

export const getPocketBase = (cookie) => {
  const userPb = new PocketBase(PB_URL);
  if (cookie) {
    userPb.authStore.loadFromCookie(cookie);
  }
  return userPb;
};

export const getAdminPocketbase = async () => {
  return adminPb;
};
```

The project is big so I want to maintain the functions so the imports don't have to be updated.
